File: Trees_Graphs/go/paths_with_sum.go
package main

import (
  "fmt"
  "types"
)


func countNums(node *types.TreeNode, sum int, value int, counter *int, passed map[*types.TreeNode]bool) {
  if node.LeftChild != nil {
    if sum + node.LeftChild.Value == value && !passed[node.LeftChild] {
      passed[node.LeftChild] = true
      *counter++
    }
    countNums(node.LeftChild, node.LeftChild.Value, value, counter, passed)
    countNums(node.LeftChild, sum + node.LeftChild.Value, value, counter, passed)
  }
  if node.RightChild != nil {
    if sum + node.RightChild.Value == value && !passed[node.RightChild]{
      passed[node.LeftChild] = true
      *counter++
    }
    countNums(node.RightChild, node.RightChild.Value, value, counter, passed)
    countNums(node.RightChild, sum + node.RightChild.Value, value, counter, passed)
  }
}


func getNums(root *types.TreeNode, value int) int {
  var counter int
  passed := make(map[*types.TreeNode]bool)

  countNums(root, root.Value, value, &counter, passed)

  return counter
}


func main() {
  array := []int{12, -6, 4, 3, 2, 3, 8, 1, 3, 1, 12, 6, -14, 1, 7}
  root := types.CreateTree(array)
  types.PrintTreeNodes(root)

  fmt.Println(getNums(root, 6))
}

File: Trees_Graphs/python/paths_with_sum.py
from Tree import Tree


def count_nums(node, sum, value, passed):
    counter = 0

    if node.left_child:
        if sum + node.left_child.value == value and not passed.get(node.left_child):
            passed[node.left_child] = True
            counter += 1

        counter += count_nums(node.left_child, node.left_child.value, value, passed)
        counter += count_nums(node.left_child, sum + node.left_child.value, value, passed)

    if node.right_child:
        if sum + node.right_child.value == value and not passed.get(node.right_child):
            passed[node.right_child] = True
            counter += 1

        counter += count_nums(node.right_child, node.right_child.value, value, passed)
        counter += count_nums(node.right_child, sum + node.right_child.value, value, passed)

    return counter


def get_nums(tree, value):
    passed = {}

    return count_nums(tree.root, tree.root.value, value, passed)


def main():
    array = list(range(7))
    tree = Tree(array)
    tree.print_tree()

    print(get_nums(tree, 6))


if __name__ == '__main__':
    main()
